text/0000-read-buf.md

```diff
@@ -0,0 +1,624 @@
+2. It needs to be *efficiently* backwards compatible. Code that doesn't write `unsafe` should not be penalized by the
+    new APIs. For example, code working with a reader written before these new APIs were introduced should not become
+    slower once that code starts trying to use the new APIs.
+3. It must be compatible with `dyn Read`. Trait objects are used pervasively in IO code, so a solution can't depend on
+    monomorphization or specialization.
+4. It needs to work with both normal and vectored IO (via `read_vectored`).
+5. It needs to be composable. Readers are very commonly nested (e.g. `GzipReader<TlsStream<TcpStream>>`), and wrapper
+    readers should be able to opt-in to fast paths supported by their inner reader.
+6. A reader that does want to work directly with uninitialized memory does, at some reasonable point, need to write the
+    word `unsafe`.
+
+This RFC covers the proposed solution. For in-depth coverage of other options and the rationale for this particular
+approach over others, please refer to this [Dropbox Paper writeup](https://paper.dropbox.com/doc/IO-Buffer-Initialization--Ax97Yz2_GUH23hVjfDf4JhCAAQ-MvytTgjIOTNpJAS6Mvw38)
+or my [discussion with Niko Matsakis](http://smallcultfollowing.com/babysteps/blog/2020/01/20/async-interview-5-steven-fackler/).
+
+The proposal in the Dropbox Paper does differ from the proposal in this RFC in one significant way: its definition of
+`read_buf` returns an `io::Result<usize>` like `read` does, and the `ReadBuf` only tracks the initialized region and not
+the written-to region:
+
+```rust
+pub trait Read {
+    fn read_buf(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<usize> { ... }
+}
+```
+
+This has a subtle but important drawback. From the perspective of code working with a `Read` implementation, the
+initialization state of the buffer can be trusted to be correct, but the number of bytes read cannot! This mix of
+trusted and untrusted information can be quite a footgun for unsafe code working with a reader. For example,
+`read_to_end` needs to remember to assert that the number of bytes read is less than the number of bytes initialized
+before calling `set_len` on the `Vec<u8>` that it's reading into. Moving that bit of state into `ReadBuf` avoids the
+issue by allowing `ReadBuf` to guarantee that these two values stay consistent.
+
+# Prior art
+[prior-art]: #prior-art
+
+The standard library currently has the concept of a buffer "initializer". The `Read` trait has an (unstable) method
+which returns an `Initializer` object which can take a `&mut [u8]` of uninitialized memory and initialize it as needed
+for use with the associated reader. Then the buffer is just passed to `read` as normal.
+
+The [`tokio::io::AsyncRead`](https://docs.rs/tokio/0.2.21/tokio/io/trait.AsyncRead.html) trait has a somewhat similar
+approach, with a `prepare_uninitialized_buffer` method which takes a `&mut [MaybeUninit<u8>]` slice and initializes it
+if necessary.
```

Comment:
It might be nice to explicitly state that it's believed that this RFC *could* replace this usage (even if the crate itself decided not to for whatever reason).